### ISSUE TITLE
ASDS sharing port

### ISSUE THREAD

**Issue by JeffreyDevloo** [openvstorage/framework-alba-plugin#227]
## Problem description
During alba alba-bench test I found that 
```
Disqualifying osd 9: Asd_client.BadLongId("qbASQD1EMz8QEAcOf1XE73CUbUazWurE", "ud1S5c7xLnUFXvDydk86lZwYDJFLwYR1")
```
was occuring for every extra asd that I specified.
The flow that I used was: 
- init disks and claim asds
- run alba bench
- remove all disks and increment the asds count
- back to step 1

Further investigation lead to:

```
12 : { Nsm_model.OsdInfo.kind =  Nsm_model.OsdInfo.Asd ((["172.22.1.30"], 8618, false, false),    "jS3qR7sAlCqvN2gF3tzw0UeJP11eDcSf");  decommissioned = false; node_id = "9JnUfmlqdQOzwLLLhOk3KcvXrjhyrMOW";  other =  "{ \"id\" : \"jS3qR7sAlCqvN2gF3tzw0UeJP11eDcSf\", \"node_id\" : \"9JnUfmlqdQOzwLLLhOk3KcvXrjhyrMOW\", \"port\" : 8618, \"used_bytes\" : \"0\", \"total_bytes\" : \"239933292544\", \"version\" : \"AsdV1\", \"network_interfaces\":[{\"ipv4_addr\": \"172.22.1.30\"} ]}";  total = 239933292544L; used = 0L; seen = [2016/10/04_12:46:19.5419];  read = []; write = []; errors = [] }57 : { Nsm_model.OsdInfo.kind =  Nsm_model.OsdInfo.Asd ((["172.22.1.30"], 8618, false, false),    "kY7ipX4Kv1KZ7wDgMIBZoOFTLLB0qsCv");  decommissioned = false; node_id = "9JnUfmlqdQOzwLLLhOk3KcvXrjhyrMOW";  other =  "{ \"id\" : \"kY7ipX4Kv1KZ7wDgMIBZoOFTLLB0qsCv\", \"node_id\" : \"9JnUfmlqdQOzwLLLhOk3KcvXrjhyrMOW\", \"port\" : 8618, \"used_bytes\" : \"0\", \"total_bytes\" : \"239933292544\", \"version\" : \"AsdV1\", \"network_interfaces\":[{\"ipv4_addr\": \"172.22.1.30\"} ]}";  total = 239933292544L; used = 0L; seen = [2016/10/04_13:09:44.5317];  read = []; write = []; errors = [] }
```
Summarized: the asds are using the same port.

### Setup

Hyperconverged setup

- Two nodes where one had 6 disks to initialize

### Package information

- openvstorage                          2.7.3-rev.4059.6d404af-1                                 amd64        openvStorage
- openvstorage-backend                  1.7.3-rev.721.b4724e5-1                                  amd64        openvStorage Backend plugin
- openvstorage-backend-core             1.7.3-rev.721.b4724e5-1                                  amd64        openvStorage Backend plugin core
- openvstorage-backend-webapps          1.7.3-rev.721.b4724e5-1                                  amd64        openvStorage Backend plugin Web Applications
- openvstorage-cinder-plugin            1.2.2-rev.38.dcc3b76-1                                   amd64        OpenvStorage Cinder plugin for OpenStack
- openvstorage-core                     2.7.3-rev.4059.6d404af-1                                 amd64        openvStorage core
- openvstorage-hc                       1.7.3-rev.721.b4724e5-1                                  amd64        openvStorage Backend plugin HyperConverged
- openvstorage-sdm                      1.6.3-rev.372.0cfe8a2-1                                  amd64        Open vStorage Backend ASD Manager
- openvstorage-webapps                  2.7.3-rev.4059.6d404af-1                                 amd64        openvStorage Web Applications

**Comment by khenderick** [comment]
After investigation with @domsj , it turned out that certain asds were not (correctly) purged with Alba. It seems the Framework code absorbs exceptions on purge, and continue to clean up the Asd. However, if a new Asd is created, it might use the same port and register it with the same Alba backend, which still knows about the previous instance. This obviously causes issues.

**Comment by domsj** [comment]
It doesn't cause too much issues though, the proxy/benchmark/maintenance will quickly disqualify it.
After 15 minutes (by default) we'll even start repairing it. Some data in the nsms will however never be cleaned up due to this.

**Comment by khenderick** [comment]
Duplicate asds:
* 172.22.1.30:8618 with id ```jS3qR7sAlCqvN2gF3tzw0UeJP11eDcSf```
* 172.22.1.30:8618 with id ```kY7ipX4Kv1KZ7wDgMIBZoOFTLLB0qsCv```

After investigation, on node ```cmp04```, I see a task to remove ```jS3...```, which completes about 32 seconds later. While our code ignores exceptions on ```purge-osd``` (which should be removed) it does log the error. However, no error is found indicating that the ```purge-osd``` was executed without any issues.

The only thing I can do here is adding some more logging and make sure that an execption during ```purge-osd``` is not swallowed.

**Comment by khenderick** [comment]
Added logging by #228, packaged in openvstorage-backend-1.7.4-rev.727.349f39c. Please try again, and if the same issue pops up, we'll have some more logging.

**Comment by khenderick** [comment]
Got more logging which revealed that as soon as 2 asds are claimed on a single disk, some things get added to the model incorrectly, causing the above issue, but also causing the backend to be unremovable.

**Comment by khenderick** [comment]
Fixed by #229, packaged in openvstorage-backend-1.7.4-rev.729.69c9518

**Comment by JeffreyDevloo** [comment]
### Steps
The following steps were automated during a benchmark test:
- Installed hyperconverged cluster
- Assigned disk roles
- Added a backend
- Initialized disks
- Claimed asds
- Benchmark
- Removed asds
- Removed disks
- Go back to initialized disks and repeat until asd/ssd is 16

### Output
Asd removal tasks are successful: 
- Purge record in the log
- Successful deletion record
```
2016-10-11 10:04:23 27400 +0200 - cmp01 - 21749/140494218741568 - celery/celery.worker.strategy - 7743 - INFO - Received task: albanode.reset_asd[3a1c62ce-1671-4d19-974e-40ab73f45a09]
2016-10-11 10:04:23 27400 +0200 - cmp01 - 21749/140494218741568 - celery/celery.pool - 7744 - DEBUG - TaskPool: Apply <function _fast_trace_task at 0x7fc7568bbc80> (args:('albanode.reset_asd', '3a1c62ce-1671-4d19-974e-40ab73f45a09', ('e589df0b-751e-41c7-a624-4178ffd06db8', u'pMhaYnVigdcfxbk0kXf3B08GFHATwakt', {u'good': 1, u'critical': 0, u'lost': 0}), {}, {'utc': True, u'is_eager': False, 'chord': None, u'group': None, 'args': ('e589df0b-751e-41c7-a624-4178ffd06db8', u'pMhaYnVigdcfxbk0kXf3B08GFHATwakt', {u'good': 1, u'critical': 0, u'lost': 0}), 'retries': 0, u'delivery_info': {u'priority': 0, u'redelivered': False, u'routing_key': u'generic.#', u'exchange': u'generic'}, 'expires': None, u'hostname': 'celery@cmp01', 'task': 'albanode.reset_asd', 'callbacks': None, u'correlation_id': u'3a1c62ce-1671-4d19-974e-40ab73f45a09', 'errbacks': None, 'timelimit': (None, None), 'taskset': None, 'kwargs': {}, 'eta': None, u'reply_to': u'bcba57af-fbc4-3ea5-8a30-89e0b5f16d45', 'id': '3a1c62ce-1671-4d19-974e-40ab73f45a09', u'headers': {}}) kwargs:{})
2016-10-11 10:04:23 27400 +0200 - cmp01 - 21749/140494218741568 - celery/celery.worker.job - 7745 - DEBUG - Task accepted: albanode.reset_asd[3a1c62ce-1671-4d19-974e-40ab73f45a09] pid:8597
2016-10-11 10:04:23 27900 +0200 - cmp01 - 8597/140494218741568 - lib/scheduled tasks - 7740 - INFO - Ensure single CHAINED mode - ID 1476173063_y3CzmPtLp5 - Amount of jobs pending for key ovs_ensure_single_albanode.remove_asd: 0
2016-10-11 10:04:23 28000 +0200 - cmp01 - 8597/140494218741568 - lib/scheduled tasks - 7741 - INFO - Ensure single CHAINED mode - ID 1476173063_y3CzmPtLp5 - New task albanode.remove_asd with params {'node_guid': 'e589df0b-751e-41c7-a624-4178ffd06db8', 'asd_id': u'pMhaYnVigdcfxbk0kXf3B08GFHATwakt', 'expected_safety': {u'good': 1, u'critical': 0, u'lost': 0}} scheduled for execution
2016-10-11 10:04:23 28100 +0200 - cmp01 - 8597/140494218741568 - lib/scheduled tasks - 7742 - INFO - Ensure single CHAINED mode - ID 1476173063_y3CzmPtLp5 - Amount of jobs pending for key ovs_ensure_single_albanode.remove_asd: 1
2016-10-11 10:04:23 28100 +0200 - cmp01 - 8597/140494218741568 - lib/scheduled tasks - 7743 - INFO - Ensure single CHAINED mode - ID 1476173063_y3CzmPtLp5 -   KWARGS: {'node_guid': 'e589df0b-751e-41c7-a624-4178ffd06db8', 'asd_id': u'pMhaYnVigdcfxbk0kXf3B08GFHATwakt', 'expected_safety': {u'good': 1, u'critical': 0, u'lost': 0}}
2016-10-11 10:04:23 28400 +0200 - cmp01 - 8597/140494218741568 - lib/albanode - 7744 - DEBUG - Removing ASD pMhaYnVigdcfxbk0kXf3B08GFHATwakt at node 172.19.5.30
2016-10-11 10:04:23 50000 +0200 - cmp01 - 8597/140494218741568 - lib/albanode - 7745 - DEBUG - Purging ASD pMhaYnVigdcfxbk0kXf3B08GFHATwakt on backend 5cee3299-1cb1-4e22-9f4a-dd148d637715
2016-10-11 10:04:23 53500 +0200 - cmp01 - 8597/140494218741568 - lib/albanode - 7746 - DEBUG - Removing ASD pMhaYnVigdcfxbk0kXf3B08GFHATwakt from disk ata-SAMSUNG_MZ7LM480HCHP-00003_S1YJNX0H801485
2016-10-11 10:04:24 91300 +0200 - cmp01 - 8597/140494218741568 - celery/celery.redirected - 7748 - WARNING - 2016-10-11 10:04:24 91300 +0200 - cmp01 - 8597/140494218741568 - extensions/asdmanagerclient - 7747 - INFO - Request "delete_asd" took 1.33 seconds (internal duration 1.32 seconds)
```

### Test result
Test passed.
